components/net/connector.rs

```diff
@@ -86,7 +95,222 @@ impl Service<Destination> for ServoHttpConnector {
     }
 }
 
-pub type Connector = HyperRustlsHttpsConnector<ServoHttpConnector>;
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+#[derive(Clone)]
+pub struct InstrumentedConnector<T> {
+    inner: HyperRustlsHttpsConnector<T>,
+}
+
+impl<T> InstrumentedConnector<T> {
+    fn new(inner: HyperRustlsHttpsConnector<T>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<T> From<HyperRustlsHttpsConnector<T>> for InstrumentedConnector<T> {
+    fn from(inner: HyperRustlsHttpsConnector<T>) -> Self {
+        Self::new(inner)
+    }
+}
+
+pub struct InstrumentedStream<T> {
+    inner: MaybeHttpsStream<T>,
+    tls_info: Option<TlsHandshakeInfo>,
+}
+
+impl<T: Unpin> Unpin for InstrumentedStream<T> {}
+
+impl<T> fmt::Debug for InstrumentedStream<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("InstrumentedStream")
+            .field("tls_info", &self.tls_info)
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct TlsHandshakeInfo {
+    pub protocol_version: Option<String>,
+    pub cipher_suite: Option<String>,
+    pub kea_group_name: Option<String>,
+    pub signature_scheme_name: Option<String>,
+    pub alpn_protocol: Option<String>,
+    pub certificate_chain_der: Vec<Vec<u8>>,
+    pub used_ech: bool,
+}
+
+impl TlsHandshakeInfo {
+    fn from_connection(conn: &ClientConnection) -> Self {
+        let protocol_version = conn
+            .protocol_version()
+            .map(protocol_version_to_string);
+        let cipher_suite = conn
+            .negotiated_cipher_suite()
+            .map(|suite| format!("{:?}", suite.suite()));
+        let kea_group_name = conn
+            .negotiated_key_exchange_group()
+            .map(|group| format!("{:?}", group.name()));
+        let certificate_chain_der = conn
+            .peer_certificates()
+            .map(|certs| certs.iter().map(|cert| cert.as_ref().to_vec()).collect())
+            .unwrap_or_default();
+        let alpn_protocol = conn
+            .alpn_protocol()
+            .map(|proto| String::from_utf8_lossy(proto).into_owned());
+        let used_ech = matches!(conn.ech_status(), EchStatus::Accepted);
+
+        Self {
+            protocol_version,
+            cipher_suite,
+            kea_group_name,
+            signature_scheme_name: None,
+            alpn_protocol,
+            certificate_chain_der,
+            used_ech,
+        }
+    }
+}
+
+fn protocol_version_to_string(version: ProtocolVersion) -> String {
+    match version {
+        ProtocolVersion::TLSv1_3 => "TLS 1.3".to_string(),
+        ProtocolVersion::TLSv1_2 => "TLS 1.2".to_string(),
+        ProtocolVersion::TLSv1_1 => "TLS 1.1".to_string(),
+        ProtocolVersion::TLSv1_0 => "TLS 1.0".to_string(),
+        ProtocolVersion::SSLv3 => "SSL 3.0".to_string(),
+        ProtocolVersion::DTLSv1_2 => "DTLS 1.2".to_string(),
+        ProtocolVersion::DTLSv1_0 => "DTLS 1.0".to_string(),
+        other => format!("{other:?}"),
```

Comment:
Can we use an exhaustive match here, so that we don't accidentally ignore any new variants that are added in the future?